src/SFA.DAS.ApprenticeFeedback.Application/Queries/GetEmailTransactions/GetEmailTransactionsQueryHandler.cs

```diff
@@ -0,0 +1,65 @@
+﻿
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using MediatR;
+using SFA.DAS.ApprenticeFeedback.Domain.Configuration;
+using SFA.DAS.ApprenticeFeedback.Domain.Entities;
+using SFA.DAS.ApprenticeFeedback.Domain.Interfaces;
+using static SFA.DAS.ApprenticeFeedback.Domain.Models.Enums;
+
+namespace SFA.DAS.ApprenticeFeedback.Application.Queries.GetEmailTransactions
+{
+    public class GetEmailTransactionsQueryHandler : IRequestHandler<GetEmailTransactionsQuery, GetEmailTransactionsResult>
+    {
+        private readonly IFeedbackTransactionContext _transactionContext;
+        private readonly IApprenticeFeedbackTargetContext _targetContext;
+        private readonly ApplicationSettings _settings;
+
+        public GetEmailTransactionsQueryHandler(IFeedbackTransactionContext transactionContext, IApprenticeFeedbackTargetContext targetContext, ApplicationSettings settings)
+        {
+            _transactionContext = transactionContext;
+            _targetContext = targetContext;
+            _settings = settings;
+        }
+
+        public async Task<GetEmailTransactionsResult> Handle(GetEmailTransactionsQuery request, CancellationToken cancellationToken)
+        {
+            DateTime createdOn = DateTime.Now;
+
+            IEnumerable<Guid?> ineligibleTargetIds = await
```

Comment:
I think this may end up being a performance bottleneck. We will end up with thousands of transaction records, along with thousands of apprentice feedback targets.

I was under the impression the original plan was to do a Merge statement via a stored proc to allow the database to do the heavy lifting so that the command handler can say "generate me X transactions please"

Reply:
Had a bit of a chat with Al on this one, think it may be easier to drop the application configuration limit of how many to create in favour of a stored proc that just adds in any missing transactions:
Below is what Al came up with.
```INSERT INTO [dbo].[FeedbackTransaction] ([ApprenticeFeedbackTargetId], [CreatedOn])
SELECT aft.[Id], GETUTCDATE()
FROM [dbo].[ApprenticeFeedbackTarget] aft
LEFT JOIN (SELECT [ApprenticeFeedbackTargetId] 
           FROM [dbo].[FeedbackTransaction] 
           WHERE SentDate IS NULL OR (SentDate IS NOT NULL AND SentDate >= DATEADD(day,-90,GETDATE())) ) ft1
    on ft1.[ApprenticeFeedbackTargetId] = aft.[Id]
WHERE [Status] = 2 -- "active"
AND [FeedbackEligibility] = 1 -- "allow"
AND ft1.ApprenticeFeedbackTargetId IS NULL -- no feedback, or no feedback after time delay since last send```

Reply:
He did tweak the table definition slightly as well

`CREATE TABLE [dbo].[FeedbackTransaction]
(
	-- Update Id to a sequence
	[Id] int NOT NULL IDENTITY PRIMARY KEY, 
	[ApprenticeFeedbackTargetId] UNIQUEIDENTIFIER, 
	[EmailAddress] NVARCHAR(256),
	[FirstName] NVARCHAR(256),
	[TemplateId] UNIQUEIDENTIFIER,
	[CreatedOn] DATETIME NOT NULL,
	[SendAfter] DATETIME,
	[SentDate] DATETIME
)

GO

CREATE NONCLUSTERED INDEX [IX_FeedbackTransaction_ApprenticeFeedbackTarget]
    ON [dbo].[FeedbackTransaction]	( [ApprenticeFeedbackTargetId] )
	INCLUDE ( [SentDate] );

GO`

Reply:
Changed table to reflect the above.
As mentioned it isn't possible to join all the data in multiple tables the way we are using EF (only subsets of the data), so I had to implement this as a stored procedure as there is no other way to achieve it.

Reply:
Can we also amend the length of the EmailAddress and FirstName to 200 as planned for QF-530.

Reply:
And the "90" days windows whould be a configurable parameter to be passed to the Stored proc to allow the length of time between email sends to be managed by configuration.

Reply:
Re the email address size, 254 characters I believe is the maximum, so 256 I think should stay on that basis?

Reply:
A previous comment with the stored proc above suggested dropping the 90 day (configurabe) parameter?

Reply:
Have updated to 200 as this was specified and passed PR in QF-530. Let me know if needs increasing.

Reply:
Email field is nvarchar(200) in Commitments Db where the value is originally stored on AS. 
The max length of an email address is actually 254 characters - so we could allow for this as Chris says.
However, as the email address reaches feedback via Apprentcie Commintents Registration table it is actually constrained to nvarchar(150) - but this should be at least 200 to accommodate any email addresses passed from Commitments. Interestingly, the Email field in Apprentcie Account Db is also 200.

Reply:
> A previous comment with the stored proc above suggested dropping the 90 day (configurabe) parameter?

Ian my SQL was just a sample to show how we can populate the FeedbackTransaction table with an INSERT, and wasn't supposed to be used without putting it in context.

Reply:
OK, so for the purposes of clarity, I'll change it to:
```
[EmailAddress] NVARCHAR(256),
[FirstName] NVARCHAR(200),
```
Is this right?

Reply:
OK, done